mantidproject/mantid
# Update FindUBUsingLatticeParameters for small numbers of peaks

Pascal has been having problems with this algorithm when using a very small number of peaks. Technically we should be able to use as few as 3 peaks but found the algorithm often failed after UB optimisation. Here we do two things:

1. Remove the additional UB optimisation in the `FindUBUsingLatticeParameters` algorithm. This is already performed in the  `Find_UB`, but in try catch. Now if no improvement is found, a UB is still returned.
2. A new option called `FixParameters` (default false) has been added as requested by Pascal. When this option is used the algorithm will find a UB but will not attempt to optimise using indexed HKL and Q values.

**To test:**
1. Check the tests pass
2. Try to find a UB on some data. I've provided two peaks workspaces ([peaks_data.zip](https://github.com/mantidproject/mantid/files/610891/peaks_data.zip)). One with 3 peaks and one with 4. You can use the following script to try out the changes:

```python
import os.path
path = '/Users/samueljackson/Desktop/'
peaks = LoadIsawPeaks(Filename=os.path.join(path, 'WISH00035920.peaks'))
FindUBUsingLatticeParameters(PeaksWorkspace=peaks, a=3.89, b=3.89, c=20.69, alpha=90, beta=90, gamma=90)
FindUBUsingLatticeParameters(PeaksWorkspace=peaks, a=3.89, b=3.89, c=20.69, alpha=90, beta=90, gamma=90, FixParameters=True)

peaks2 = LoadIsawPeaks(Filename=os.path.join(path, 'WISH00035920_2.peaks'))
FindUBUsingLatticeParameters(PeaksWorkspace=peaks2, a=3.89, b=3.89, c=20.69, alpha=90, beta=90, gamma=90)
FindUBUsingLatticeParameters(PeaksWorkspace=peaks2, a=3.89, b=3.89, c=20.69, alpha=90, beta=90, gamma=90, FixParameters=True)
````
You should see output listed similar to the below. Note that all of these return a UB despite the low number of peaks (in the case of the first workspace only 2 out of three are indexed).

```
LoadIsawPeaks started
LoadIsawPeaks successful, Duration 0.50 seconds
FindUBUsingLatticeParameters started
Error = 0.0176354
UB = Matrix(3,3)0.0573513,0.198464,-0.0287652,0.244488,-0.00988809,0.0148194,0.0549668,-0.163092,-0.0359024
New UB will index 2 Peaks out of 3 with tolerance 0.150
Lattice Parameters:    3.890000    3.890000   20.690000   90.000000   90.000000   90.000000   313.083149
Lattice Parameters (Refined - Input):   -0.000000   -0.000000    0.000000    0.000000    0.000000    0.000000
FindUBUsingLatticeParameters successful, Duration 1.10 seconds
FindUBUsingLatticeParameters started
Error = 0.0176354
UB = Matrix(3,3)0.0573513,0.198464,-0.0287652,0.244488,-0.00988809,0.0148194,0.0549668,-0.163092,-0.0359024
New UB will index 2 Peaks out of 3 with tolerance 0.150
Lattice Parameters:    3.890000    3.890000   20.690000   90.000000   90.000000   90.000000   313.083149
Lattice Parameters (Refined - Input):   -0.000000   -0.000000    0.000000    0.000000    0.000000    0.000000
FindUBUsingLatticeParameters successful, Duration 1.14 seconds
LoadIsawPeaks started
LoadIsawPeaks successful, Duration 0.54 seconds
FindUBUsingLatticeParameters started
Error = 3.74849e-30
UB = Matrix(3,3)0.177645,0.219583,0.000918923,0.0515916,0.00840966,0.0499153,0.244531,-0.138228,0.0010567
New UB will index 3 Peaks out of 4 with tolerance 0.150
Lattice Parameters:    3.332754    3.862780   20.426618   91.336109  101.171713   93.725987   257.264496
Lattice Parameters (Refined - Input):   -0.557246   -0.027220   -0.263382    1.336109   11.171713    3.725987
FindUBUsingLatticeParameters successful, Duration 1.11 seconds
FindUBUsingLatticeParameters started
Error = 0.0261459
UB = Matrix(3,3)0.141237,0.212968,-0.00525575,0.0402145,0.00664644,0.0477211,0.210997,-0.143823,-0.00557722
New UB will index 3 Peaks out of 4 with tolerance 0.150
Lattice Parameters:    3.890000    3.890000   20.690000   90.000000   90.000000   90.000000   313.083149
Lattice Parameters (Refined - Input):    0.000000    0.000000    0.000000    0.000000    0.000000    0.000000
FindUBUsingLatticeParameters successful, Duration 1.18 seconds
```

Fixes #18039 .

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####


##### Functional Tests #####


- Does everything look good? Comment with the ship it emoji but don't merge. A member of `@mantidproject/gatekeepers` will take care of it.
